cfme/tests/physical_infrastructure/test_redfish_physical_server_details.py

```diff
@@ -20,3 +33,45 @@ def test_redfish_physical_server_details_stats(physical_server):
         initialEstimate: None
     """
     physical_server.validate_stats(ui=True)
+
+
+# Power Button
+def test_redfish_server_power_off(physical_server, provider):
```

Comment:
This appears to be setup to assume that this specific set of tests is executing in sequence?  Not an acceptable pattern if so, as there are too many failure modes introduced. We cannot guarantee that these tests will execute on the same appliance, by the same pytest slave, much less in sequence.   What if I just decided to run the power_on test in isolation - the vm is already powered on, did I really test anything?

If it is assumed that the machine is in a powered on state when this test runs, that needs to be built into this test function, or into a fixture used by this test.

Same goes for the following tests that may be assuming a specific power state at the start of the test.

Few things, will enumerate:
1. Docblocks for these test methods, explaining any assumptions about the server state, or whether the state matters
2. Consider parametrizing some here to reduce the number of methods, maybe a single `test_redfish_server_power_control` test that goes through the power operations in turn.
3. Some assertion that the server actually went into the requested state? Currently this is just testing that I can make that button click in MIQ, but not that it has any effect on the server.  Look at using the wrapanapi classes/methods for this power state assertion.

Reply:
Now that I said all of this, I see this thread:
https://github.com/ManageIQ/integration_tests/pull/8236#discussion_r242804860

So I would then boil down my issues to:
1. The assumptions made here (all power operations are available always in UI) need to be in the test docblock
2. There is no need to have these separated into individual test functions, this needs to be parametrized.  Consider a namedTuple of button + flash message, and parametrize over a list of tuples for a single `test_redfish_power_buttons` test function.

There are also assumptions being made in calling `server.power_off` or `power_on` functions, in that they are using a UI context to click some button.  If the test is just that these buttons are available, and that clicking them produces flash messages, than I think the test should directly navigate to server details and use the widget methods to assert that the options are available, then click them.

Ultimately I think the test should be filled out to actually assert a power state, call a power operation, and then actually assert the power state change in addition to the flash message check.  We have power control test modules in the cloud/infra or cloud_infra_common folder, I suggest you check out how they're structured.

Reply:
The only assumption we have is that any button can be pressed at any time regardless of the server's state. The test just makes sure that we can queue a request. So the tests can be run in any order. I will write that  in the docblocks to make it clearer.

I kept the single button press in each test method because I've taken the example from [Lenovo](https://github.com/ManageIQ/integration_tests/blob/master/cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py#L31) and it seemed simple and understandable enough. Of course in the interest of reducing the number of test methods (and thus the test execution time), I can put them in a single test as you suggested. Within this test I can also do the widget tests as you suggested.

Regarding the complexity of the depth of these tests, our intent is really to test that the GUI is properly wired with the provider in that clicking a button results in a reported success on the GUI. Admittedly we currently lack the assertion that clicking a certain button actually confers the corresponding command to the back-end. But we do not want to test the physical servers' state transitions here, because this part is very Redfish-vendor specific, so results might vary depending on the server the provider instance is connected to.

Reply:
Done.